indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java

```diff
@@ -318,6 +324,114 @@ public void handle()
+        if (spec.isSuspended()) {
+          log.info("[%s] supervisor is suspended, skip to check dynamic allocate task logic", dataSource);
+          return;
+        }
+        log.debug("PendingCompletionTaskGroups is [%s] for dataSource [%s].", pendingCompletionTaskGroups, dataSource);
+        for (CopyOnWriteArrayList list : pendingCompletionTaskGroups.values()) {
+          if (!list.isEmpty()) {
+            log.info("Still hand off tasks unfinished, skip to do scale action [%s] for dataSource [%s].", pendingCompletionTaskGroups, dataSource);
+            return;
+          }
+        }
+        if (autoScalerConfig != null && nowTime - dynamicTriggerLastRunTime < autoScalerConfig.getMinTriggerDynamicFrequencyMillis()) {
+          log.info("NowTime - dynamicTriggerLastRunTime is [%s]. Defined minTriggerDynamicFrequency is [%s] for dataSource [%s], CLAM DOWN NOW !", nowTime - dynamicTriggerLastRunTime, autoScalerConfig.getMinTriggerDynamicFrequencyMillis(), dataSource);
+          return;
+        }
+
+        Integer desriedTaskCount = scaleAction.call();
+        boolean allocationSuccess = dynamicAllocate(desriedTaskCount);
+
+        if (allocationSuccess) {
+          dynamicTriggerLastRunTime = nowTime;
+        }
+      }
+      catch (Exception ex) {
+        log.warn(ex, "Error, when parse DynamicAllocationTasksNotice");
+      }
+    }
+  }
+
+  /**
+   * This method determines how to do scale actions based on collected lag points.
+   * If scale action is triggered :
+   *    First of all, call gracefulShutdownInternal() which will change the state of  current datasource ingest tasks from reading to publishing.
+   *    Secondly, clear all the stateful data structures: activelyReadingTaskGroups, partitionGroups, partitionOffsets, pendingCompletionTaskGroups, partitionIds. These structures will be rebuiled next 'RunNotice'.
+   *    Finally, change taskCount in SeekableStreamSupervisorIOConfig and sync it to MetaStorage.
+   * After changed taskCount in SeekableStreamSupervisorIOConfig, next RunNotice will ceate scaled number of ingest tasks without resubmitting supervisors.
+   * @param desireActiveTaskCount desire taskCount compute from autoscaler
+   * @return Boolean flag, do scale action successfully or not. If true , it will take at least 'minTriggerDynamicFrequency' before next 'dynamicAllocatie'.
+   *         If false, it will do 'dynamicAllocate' again after 'dynamicCheckPeriod'.
+   * @throws InterruptedException
+   * @throws ExecutionException
+   * @throws TimeoutException
+   */
+  private boolean dynamicAllocate(Integer desireActiveTaskCount) throws InterruptedException, ExecutionException, TimeoutException
+  {
+    int currentActiveTaskCount;
+    Collection<TaskGroup> activeTaskGroups = activelyReadingTaskGroups.values();
+    currentActiveTaskCount = activeTaskGroups.size();
+
+    if (desireActiveTaskCount == -1) {
+      return false;
+    } else {
```

Comment:
Nice catch! It is necessary to do this check. Thanks.

Reply:
Actually desireActiveTaskCount can't be equal to currentActiveTaskCount here. Take scale out as an example :
Here is the logic of computing desireActiveTaskCount in autoscaler
```
      if (currentActiveTaskCount == defaultAutoScalerConfig.getTaskCountMax()) {
        log.info("CurrentActiveTaskCount reach task count Max limit, skip to scale out tasks for dataSource [%s].", dataSource);
        return -1;
      } else {
        desiredActiveTaskCount = Math.min(taskCount, defaultAutoScalerConfig.getTaskCountMax());
      }
```
Condition 1: `int taskCount = currentActiveTaskCount + defaultAutoScalerConfig.getScaleOutStep()`
Condition 2: `currentActiveTaskCount != defaultAutoScalerConfig.getTaskCountMax()` otherwise will return -1.
Condition 3: `desiredActiveTaskCount = Math.min(taskCount, defaultAutoScalerConfig.getTaskCountMax())`
Condition 4: `defaultAutoScalerConfig.getScaleOutStep() > 0`

=>  get **Condition5** based on Condition 1 and Condition 3: `desiredActiveTaskCount = Math.min(currentActiveTaskCount + defaultAutoScalerConfig.getScaleOutStep(), defaultAutoScalerConfig.getTaskCountMax())`

Assume `desireActiveTaskCount == currentActiveTaskCount`

Combine with Condition 4, 5 and assumption => we can get `desiredActiveTaskCount == defaultAutoScalerConfig.getTaskCountMax()`

Then based on Condition 2 => We can get `currentActiveTaskCount != desiredActiveTaskCount`
 
This conclusion conflicts with the assumption, so the assumption does not hold.
So that **desireActiveTaskCount can't be equal to currentActiveTaskCount here**

UNLESS! Users set defaultAutoScalerConfig.getScaleOutStep() = 0 by mistake.

So I changed the code as you suggested to avoid this scenario.
Thanks for your suggestion!